Babel 6

tgriesser opened the pull request:
I decided to start a fresh PR for converting to Babel 6, since #1802 seems to be a little stale. I've got the docs working in this branch, and all of the tests should be passing.

Things this PR includes:
- Babel 6
- Updated React Router
- Functioning Docs
- Removed decorator use in tests
- Some tweaks removing dead code / extraneous exports in `buildBabel.js`
- ES6 Build

The "ES6 Build" is like [react-router's](https://github.com/reactjs/react-router/blob/f0a7d1410923f90c1cd998d4bd50fee7e024d412/package.json#L13), building the source files except for any `import` / `export` statements. This, combined with a `"jsnext:main"` entry in `package.json`, and a ES6 module aware bundler like rollup or webpack 2 can significantly decrease build size.

Also, a few of the tests are slightly tweaked because it appeared that React was not `console.error`'ing multiple times for the same error message, so a test would pass in isolation but  not when the same `console.error` had been emitted in a prior test in the same run.

taion commented:
Two quick notes (I haven't read through this PR in detail yet):

- We'll want to use `module` in addition to `jsnext:main` – it's what webpack 2 supports, and Rollup now supports it as well https://github.com/rollup/rollup-plugin-node-resolve/issues/48. Node will probably never implement this, but it's pretty valuable for web stuff.
- Because of what we do with the `bsStyles` bits, &c. in `bootStrapUtils`, tree-shaking won't actually do anything useful with React-Bootstrap. The best way I'm aware of for nicely using the terser import style is `babel-plugin-lodash`, but I'm concerned it might just confuse people if we talk about it in React-Bootstrap docs.

taion commented:
Looks like tools/es6 didn't get checked in. Probably want the `.gitignore` to hit `/es6` instead of `es6`.

Also, it should probably just be `/es` instead of `/es6`. At least, that's what Redux uses, and it's what I cut the React Router alpha to use as well.

tgriesser commented:
> The best way I'm aware of for nicely using the terser import style is babel-plugin-lodash

I've been using that as well and it's been really nice. Agreed that it might be confusing to add in the docs, folks who know what to look for will find it on their own. I was thinking this might also be a nice plugin to use internally here to clean up the need for full paths for `react-overlays/lib/*` imports.

The other thing I wanted to propose but thought too much to include in this PR was to add `src` to the files list in `package.json`. This would allow those who wanted to modify their webpack settings to build directly from the `src`, enabling them to configure their own babel parsing presets. 

For example this would be a good solution to #2101, by allowing you to optionally add [`react-remove-prop-types`](https://github.com/oliviertassinari/babel-plugin-transform-react-remove-prop-types) to the build process, without any additional build steps for this project.

----

> We'll want to use module in addition to jsnext:main

Added

> Looks like tools/es6 didn't get checked in

Fixed

> Also, it should probably just be /es instead of /es6

Updated

taion commented on .babelrc:
```diff
@@ -1,6 +1,9 @@
-  "stage": 1,
-  "optional": ["runtime"],
-  "loose": ["all"],
-  "plugins": ["dev-expression"]
```
Still need `babel-plugin-dev-expression`

tgriesser replied:
Ah interesting! I didn't realize this did more than replace `__DEV__`, looks pretty useful!

taion commented:
It's a bit annoying to get babel-plugin-istanbul working well. I'd just upgrade isparta loader to v2. That's been working fine in most of my repos.

taion commented:
Huh, it just works?

tgriesser commented:
@taion just swapped it out, but yeah it seemed to work - what sort of issues were you seeing?

taion commented:
Huh maybe there weren't any. You'd still need to restrict the coverage calculation to just the contents of `src/` though, and exclude `tests/`.

taion commented:
Thanks! LGTM

tgriesser commented:
@taion Cool, I'll just stick with `isparta-loader` for now and then upgrading to `babel-plugin-istanbul` can be a separate PR

AlexKVal commented:
Great job 👍

taion commented:
Thanks! Released in v0.30.2.
